### PR TITLE
Tc1 user preference algorithm

### PR DESCRIPTION
**Pull request by Brianna-Flo**
## Description

<what this pull request is doing>

- Adjust the recommendation algorithm to allow a user to input their preferred cooking time blocks
    - Allow a user to have multiple preferred time blocks
    - Present to a user up to 2 algorithm generated options for the time slots to fit in their schedule to cook the meal
- Allow the user to select whether to cook throughout the week or schedule one event to cook all meals and present events accordingly

<what will be done in later PRs and not included here>

- UI response for error handling for user time block input

## Milestones
<the milestones or stories/features that this works towards>

- TC1, user ability to customize event recommendations to generate the most optimal recommendations

## Resources
<links to tutorials, code snippets, inspirations>
<if you took or translated specific code from the internet, please call it out here!>
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/

## Test Plan
<insert images or gifs of feature>
<otherwise, say how code was tested if not UI facing>
<any edge cases you might have specifically tested>
<div>
    <a href="https://www.loom.com/share/4319893fac0340c5975fcb22b91275fb">
      <p>Cook throughout the week: User Time Preference and Event Selection Demo - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/4319893fac0340c5975fcb22b91275fb">
      
    </a>
  </div>
<div>
    <a href="https://www.loom.com/share/d778e4a003f84a10ac827749394c83dd">
      <p>Single Prep Day: User Time Preference and Event Selection Demo - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/d778e4a003f84a10ac827749394c83dd">
      
    </a>
  </div>

### CODE REVIEW

File: backend/utils/calendarUtils.ts [hunk 1]
```diff
@@ -45,6 +45,49 @@ type GPRecipeEventTypes = {
...
+  // TODO determine heuristic to estimate cooking all recipes
+  const estimatedCookTime = longestCookTime * 2;
+  for (const freeBlock of userFreeTime) {
+    const timeOptions = fitsUserPreferences({
+      freeBlock,
+      userPreferences,
+      readyInMinutes: estimatedCookTime,
+    });
+    if (timeOptions) {
+      const bestBlock = {
```
Comment (paulshaoyuqiao) [inline backend/utils/calendarUtils.ts:73]: can we declare a data model/class over this, rather than using vanilla dictionary? 
this way we will have finer control over read-only access and add any additional operation we want to perform on the block object.

File: backend/utils/calendarUtils.ts [hunk 2]
```diff
@@ -144,42 +192,62 @@ const fitsUserPreferences = ({
...
       end = preferredEnd;
     }
-    if (end - start >= recipe.readyInMinutes) {
-      let optionArray = [{start: new Date(start),
-        end: new Date(start + recipe.readyInMinutes * 1000 * 60),}]
-      start = start + 1000 * 60 * 15
-      if (end - start >= recipe.readyInMinutes) {
-        optionArray = [...optionArray, { start: new Date(start),
-        end: new Date(start + recipe.readyInMinutes * 1000 * 60),}]
+    if (end - start >= readyInMinutes * 1000 * 60) {
```
Comment (paulshaoyuqiao) [inline backend/utils/calendarUtils.ts:195]: let's make 1000 and 60 all constants at the top

Comment (paulshaoyuqiao) [reply backend/utils/calendarUtils.ts:195]: also readyInMinutes * 1000 * 60 can probably be defined at the start so we don't have to redefining the calculation so it is easier to read.

File: backend/utils/calendarUtils.ts [hunk 3]
```diff
@@ -144,42 +192,62 @@ const fitsUserPreferences = ({
...
+      let optionArray = [
+        {
+          start: new Date(start),
+          end: new Date(start + readyInMinutes * 1000 * 60),
+        },
+      ];
+      start = start + 1000 * 60 * 15;
```
Comment (paulshaoyuqiao) [inline backend/utils/calendarUtils.ts:202]: add a comment to explain what 15 means; refactor out 15 too into a constant

**Review by paulshaoyuqiao — CHANGES_REQUESTED** [review]
